File: src/main/java/skytils/skytilsmod/utils/toasts/GuiToast.java
package skytils.skytilsmod.utils.toasts;


import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.gui.ScaledResolution;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.util.MathHelper;
import org.lwjgl.opengl.GL11;

import javax.annotation.Nullable;
import java.nio.FloatBuffer;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;

/**
 * Taken from Skyblockcatia under MIT License
 * Modified
 * https://github.com/SteveKunG/SkyBlockcatia/blob/1.8.9/LICENSE.md
 *
 * @author SteveKunG
 */
public class GuiToast extends Gui {
    protected final Minecraft mc;
    private final GuiToast.ToastInstance<?>[] visible = new GuiToast.ToastInstance[5];
    private final Deque<IToast> toastsQueue = new ArrayDeque<>();

    public GuiToast(Minecraft mc) {
        this.mc = mc;
    }

    public static void drawSubline(GuiToast toastGui, long delta, long firstDrawTime, long maxDrawTime, FloatBuffer buffer, String subLine, boolean shadow) {
        long minDraw = (long) (maxDrawTime * 0.1D);
        long maxDraw = maxDrawTime + 500L;
        long backwardDraw = (long) (maxDrawTime * 0.5D);
        long textSpeed = 1500L + (long) (maxDrawTime * 0.1D);
        int x = 30;
        int textWidth = toastGui.mc.fontRendererObj.getStringWidth(subLine);
        int maxSize = textWidth - 135;
        long timeElapsed = delta - firstDrawTime - minDraw;
        long timeElapsed2 = maxDraw - delta - backwardDraw;
        int maxTextLength = 125;

        if (textWidth > maxSize && textWidth > maxTextLength) {
            if (timeElapsed > 0) {
                x = Math.max((int) (-textWidth * timeElapsed / textSpeed + x), -maxSize + 16);
            }

            int backward = Math.max(Math.min((int) -(textWidth * timeElapsed2 / textSpeed), 30), -maxSize + 16);

            if (timeElapsed > timeElapsed2) {
                x = backward;
            }
        }

        ScaledResolution res = new ScaledResolution(toastGui.mc);
        double height = res.getScaledHeight();
        double scale = res.getScaleFactor();
        float[] trans = new float[16];

        buffer.clear();
        GL11.glGetFloat(GL11.GL_MODELVIEW_MATRIX, buffer);
        buffer.get(trans);
        float xpos = trans[12];

        GL11.glEnable(GL11.GL_SCISSOR_TEST);
        GL11.glScissor((int) ((xpos + 29) * scale), (int) ((height - 196) * scale), (int) (126 * scale), (int) (195 * scale));

        if (shadow) {
            toastGui.mc.fontRendererObj.drawStringWithShadow(subLine, x, 18, 0xFFFFFF);
        } else {
            toastGui.mc.fontRendererObj.drawString(subLine, x, 18, 0xFFFFFF);
        }

        GL11.glDisable(GL11.GL_SCISSOR_TEST);
    }

    public void drawToast(ScaledResolution resolution) {
        RenderHelper.disableStandardItemLighting();

        for (int i = 0; i < this.visible.length; ++i) {
            GuiToast.ToastInstance<?> toastinstance = this.visible[i];

            if (toastinstance != null && toastinstance.render(resolution.getScaledWidth(), i)) {
                this.visible[i] = null;
            }
            if (this.visible[i] == null && !this.toastsQueue.isEmpty()) {
                this.visible[i] = new GuiToast.ToastInstance(this.toastsQueue.removeFirst());
            }
        }
    }

    @Nullable
    public <T extends IToast> T getToast(Class<? extends T> clazz, Object obj) {
        for (GuiToast.ToastInstance<?> ins : this.visible) {
            if (ins != null && clazz.isAssignableFrom(ins.getToast().getClass()) && ins.getToast().getType().equals(obj)) {
                return (T) ins.getToast();
            }
        }
        for (IToast toast : this.toastsQueue) {
            if (clazz.isAssignableFrom(toast.getClass()) && toast.getType().equals(obj)) {
                return (T) toast;
            }
        }
        return null;
    }

    public void clear() {
        Arrays.fill(this.visible, null);
        this.toastsQueue.clear();
    }

    public boolean add(IToast toast) {
        return this.toastsQueue.add(toast);
    }

    class ToastInstance<T extends IToast> {
        private final T toast;
        private long animationTime;
        private long visibleTime;
        private IToast.Visibility visibility;

        private ToastInstance(T toast) {
            this.animationTime = -1L;
            this.visibleTime = -1L;
            this.visibility = IToast.Visibility.SHOW;
            this.toast = toast;
        }

        public T getToast() {
            return this.toast;
        }

        private float getVisibility(long delta) {
            float f = MathHelper.clamp_float((delta - this.animationTime) / 600.0F, 0.0F, 1.0F);
            f = f * f;
            return this.visibility == IToast.Visibility.HIDE ? 1.0F - f : f;
        }

        public boolean render(int x, int z) {
            long i = Minecraft.getSystemTime();

            if (this.animationTime == -1L) {
                this.animationTime = i;
            }
            if (this.visibility == IToast.Visibility.SHOW && i - this.animationTime <= 600L) {
                this.visibleTime = i;
            }

            GlStateManager.pushMatrix();
            GlStateManager.enableBlend();
            GlStateManager.enableAlpha();
            GlStateManager.tryBlendFuncSeparate(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA, 1, 0);
            GlStateManager.translate(x - 160.0F * this.getVisibility(i), z * 32, 500 + z);
            IToast.Visibility itoast$visibility = this.toast.draw(GuiToast.this, i - this.visibleTime);
            GlStateManager.disableBlend();
            GlStateManager.popMatrix();

            if (itoast$visibility != this.visibility) {
                this.animationTime = i - (int) ((1.0F - this.getVisibility(i)) * 600.0F);
                this.visibility = itoast$visibility;
            }
            return this.visibility == IToast.Visibility.HIDE && i - this.animationTime > 600L;
        }
    }
}

File: src/main/kotlin/gg/skytils/skytilsmod/utils/toasts/GuiToast.kt
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2022 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package gg.skytils.skytilsmod.utils.toasts

import gg.essential.universal.UResolution
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Gui
import net.minecraft.client.gui.ScaledResolution
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.util.MathHelper
import org.lwjgl.opengl.GL11
import java.nio.Buffer
import java.nio.FloatBuffer
import java.util.*

/**
 * Taken from Skyblockcatia under MIT License
 * Modified
 * https://github.com/SteveKunG/SkyBlockcatia/blob/1.8.9/LICENSE.md
 *
 * @author SteveKunG
 */
class GuiToast(val mc: Minecraft) : Gui() {
    private val visible = arrayOfNulls<ToastInstance<*>?>(5)
    private val toastsQueue: Deque<IToast<*>> = ArrayDeque()
    fun drawToast(resolution: ScaledResolution) {
        RenderHelper.disableStandardItemLighting()
        for (i in visible.indices) {
            val toastinstance = visible[i]
            if (toastinstance != null && toastinstance.render(resolution.scaledWidth, i)) {
                visible[i] = null
            }
            if (visible[i] == null && !toastsQueue.isEmpty()) {
                visible[i] = ToastInstance<IToast<*>>(toastsQueue.removeFirst())
            }
        }
    }

    @Suppress("unchecked_cast")
    fun <T : IToast<*>> getToast(clazz: Class<out T>, obj: Any): T? {
        for (ins in visible) {
            if (ins != null && clazz.isAssignableFrom(ins.toast!!.javaClass) && ins.toast.getType() == obj) {
                return ins.toast as T
            }
        }
        for (toast in toastsQueue) {
            if (clazz.isAssignableFrom(toast.javaClass) && toast.getType() == obj) {
                return toast as T
            }
        }
        return null
    }

    fun clear() {
        Arrays.fill(visible, null)
        toastsQueue.clear()
    }

    fun add(toast: IToast<*>): Boolean {
        return toastsQueue.add(toast)
    }

    internal inner class ToastInstance<T : IToast<*>?>(toast: T) {
        val toast: T
        private var animationTime: Long
        private var visibleTime: Long
        private var visibility: IToast.Visibility?
        private fun getVisibility(delta: Long): Float {
            var f = MathHelper.clamp_float((delta - animationTime) / 600.0f, 0.0f, 1.0f)
            f *= f
            return if (visibility == IToast.Visibility.HIDE) 1.0f - f else f
        }

        fun render(x: Int, z: Int): Boolean {
            val i = Minecraft.getSystemTime()
            if (animationTime == -1L) {
                animationTime = i
            }
            if (visibility == IToast.Visibility.SHOW && i - animationTime <= 600L) {
                visibleTime = i
            }
            GlStateManager.pushMatrix()
            GlStateManager.enableBlend()
            GlStateManager.enableAlpha()
            GlStateManager.tryBlendFuncSeparate(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA, 1, 0)
            GlStateManager.translate(x - 160.0f * getVisibility(i), (z * 32).toFloat(), (500 + z).toFloat())
            val toastVisibility = toast!!.draw(this@GuiToast, i - visibleTime)
            GlStateManager.disableBlend()
            GlStateManager.popMatrix()
            if (toastVisibility != visibility) {
                animationTime = i - ((1.0f - getVisibility(i)) * 600.0f).toInt()
                visibility = toastVisibility
            }
            return visibility == IToast.Visibility.HIDE && i - animationTime > 600L
        }

        init {
            animationTime = -1L
            visibleTime = -1L
            visibility = IToast.Visibility.SHOW
            this.toast = toast
        }
    }

    companion object {
        fun drawSubline(
            toastGui: GuiToast,
            delta: Long,
            firstDrawTime: Long,
            maxDrawTime: Long,
            buffer: FloatBuffer,
            subLine: String?,
            shadow: Boolean
        ) {
            val minDraw = (maxDrawTime * 0.1).toLong()
            val maxDraw = maxDrawTime + 500L
            val backwardDraw = (maxDrawTime * 0.5).toLong()
            val textSpeed = 1500L + (maxDrawTime * 0.1).toLong()
            var x = 30
            val textWidth = toastGui.mc.fontRendererObj.getStringWidth(subLine)
            val maxSize = textWidth - 135
            val timeElapsed = delta - firstDrawTime - minDraw
            val timeElapsed2 = maxDraw - delta - backwardDraw
            val maxTextLength = 125
            if (textWidth > maxSize && textWidth > maxTextLength) {
                if (timeElapsed > 0) {
                    x = (-textWidth * timeElapsed / textSpeed + x).toInt().coerceAtLeast(-maxSize + 16)
                }
                val backward =
                    (-(textWidth * timeElapsed2 / textSpeed)).toInt().coerceAtMost(30).coerceAtLeast(-maxSize + 16)
                if (timeElapsed > timeElapsed2) {
                    x = backward
                }
            }
            val res = UResolution
            val height = res.scaledHeight.toDouble()
            val scale = res.scaleFactor
            val trans = FloatArray(16)
            (buffer as Buffer).clear()
            GL11.glGetFloat(GL11.GL_MODELVIEW_MATRIX, buffer)
            buffer[trans]
            val xpos = trans[12]
            GL11.glEnable(GL11.GL_SCISSOR_TEST)
            GL11.glScissor(
                ((xpos + 29) * scale).toInt(),
                ((height - 196) * scale).toInt(),
                (126 * scale).toInt(),
                (195 * scale).toInt()
            )
            if (shadow) {
                toastGui.mc.fontRendererObj.drawStringWithShadow(subLine, x.toFloat(), 18f, 0xFFFFFF)
            } else {
                toastGui.mc.fontRendererObj.drawString(subLine, x, 18, 0xFFFFFF)
            }
            GL11.glDisable(GL11.GL_SCISSOR_TEST)
        }
    }
}